daemon: automatic public nodes discovering and bootstrap daemon switching

xiphon opened the pull request:
Implemented `--bootstrap-daemon-address auto` mode.

Allows to sync and use the wallet while the daemon is syncing with the network.

* **While the daemon is syncing**
it will automatically select random public node (node that is running with `--public-node` mode enabled) to serve incoming requests (i.e. will use it as a remote node). If currently selected public node fails, the daemon will switch to another randomly chosen public node.
* **Once the daemon is fully synced**
it will use the local blockchain to serve incoming requests, i.e. operating in normal mode.

erciccione commented:
Thanks for this @xiphon, i think would be a very nice feature. Could you clarify how this would work and what flags/commands are added? Would be good to have a short explanation in plain text on this PR, for better tracking and to make easier to update user and developer guides.

xiphon commented:
@erciccione 

Updated the PR description

erciccione commented:
Nice, thank you :)

selsta commented:
Should `./monerod --bootstrap-daemon-address auto --no-sync` be possible using this PR?

xiphon commented:
> Should `./monerod --bootstrap-daemon-address auto --no-sync` be possible using this PR?

Yep. Allows a user to sync and use the wallet without downloading and storing the local blockchain.
Even tested it as well, works as intended.

fuwa0529 commented:
> Yep. Allows a user to sync and use the wallet without downloading and storing the local blockchain.

That can't be right. Can you prevent these two flags from being used simultaneously?

selsta commented:
> That can't be right. Can you prevent these two flags from being used simultaneously?

Why? `--bootstrap-daemon-address auto --no-sync` is comparable to automatically connecting to a remote node.

fuwa0529 commented:
A daemon does not join a p2p network just to serve a single user, by fetching data from the network while contributing nothing back, right? These two flags, when used together, allow a user to do just that, which should be prevented.

xiphon commented:
> Can you prevent these two flags from being used simultaneously?

Is not related to this PR

vtnerd commented on src/rpc/core_rpc_server.cpp:
```diff
@@ -1509,42 +1521,36 @@ namespace cryptonote
 
-      m_should_use_bootstrap_daemon = ok && top_height + 10 < getheight_res.height;
-      MINFO((m_should_use_bootstrap_daemon ? "Using" : "Not using") << " the bootstrap daemon (our height: " << top_height << ", bootstrap daemon's height: " << (ok ? getheight_res.height : 0) << ")");
+      uint64_t top_height = m_core.get_current_blockchain_height();
```
This is now aliasing `target_height`, but its not obvious why ... was this some technique to prevent further modifications below? Why not declare `top_height` on line 1530 and drop `target_height`? This would also elminate one additional `m_core.get_current_blockchain_height()` call which has locks, etc., and isn't cheap.

xiphon replied:
Nope, i changed the logic.

Now i want to skip public nodes that are not synced with the network (having `height` less than known `target_height`).

vtnerd commented on src/rpc/bootstrap_daemon.h:
```diff
@@ -0,0 +1,184 @@
+      return m_address;
+    }
+
+    bool get_height(uint64_t *height) noexcept
```
This looks like a decent candidate for `boost::optional<uint64_t>` return value.

vtnerd commented on src/rpc/bootstrap_daemon.h:
```diff
@@ -0,0 +1,184 @@
+    std::string m_address;
+    epee::net_utils::http::http_simple_client m_http_client;
+    bool m_last_result;
+    nodetool::node_server<cryptonote::t_cryptonote_protocol_handler<cryptonote::core>> *m_p2p;
```
The lifetime/ownership of this is unfortunate (but acceptable for its only current use case), no unit tests were added, and there is little code (outside of the randomized peer selection which can still be in a separate file) ... should the bulk of this logic just be included in the `core_rpc_server.cpp`?

xiphon commented on src/rpc/bootstrap_daemon.h:
```diff
@@ -0,0 +1,184 @@
+#pragma  once
+
+#include <iterator>
+#include <list>
```
Yep, and `#include <iterator>` as well

xiphon commented on src/rpc/bootstrap_daemon.h:
```diff
@@ -0,0 +1,184 @@
+
+    bool get_height(uint64_t *height) noexcept
+    {
+      cryptonote::COMMAND_RPC_GET_HEIGHT::request getheight_req;
```
I agree

xiphon commented on src/rpc/bootstrap_daemon.h:
```diff
@@ -0,0 +1,184 @@
+  {
+  public:
+    bootstrap_daemon(nodetool::node_server<cryptonote::t_cryptonote_protocol_handler<cryptonote::core>> *p2p)
+      : m_last_result(false)
```
I might do that, though listing members with default constructors seems questionable to me. Could you elaborate a bit more on that?

xiphon commented on src/rpc/bootstrap_daemon.h:
```diff
@@ -0,0 +1,184 @@
+      std::vector<nodetool::peerlist_entry> gray_list;
+      m_p2p->get_public_peerlist(gray_list, white_list);
+
+      const nodetool::peerlist_entry *public_node = get_random_public_node(white_list);
```
> Does `get_random_public_node` need to be public?

It is private.

> Can it be put into an anonymous namespace in a cpp and used by this function solely?

Would like to keep it as-is: private class function. Will move it to cpp tought (with all the other stuff).

xiphon commented on src/rpc/bootstrap_daemon.h:
```diff
@@ -0,0 +1,184 @@
+      json_req.method = std::string(command_name.begin(), command_name.end());
+      json_req.params = out_struct;
+
+      if (!handle_result(epee::net_utils::invoke_http_json("/json_rpc", json_req, json_resp, m_http_client)))
```
>  I don't see a practical reason for duplicating this.

is a leftover from the current codebase, will replace it with `invoke_http_json_rpc` call.

https://github.com/monero-project/monero/blob/d4d5a4433d550499921d8dd5c1bacee66c7d116b/src/rpc/core_rpc_server.cpp#L1539

xiphon commented on src/rpc/bootstrap_daemon.h:
```diff
@@ -0,0 +1,184 @@
+    {
+    }
+
+    bootstrap_daemon(const std::string &address, const boost::optional<epee::net_utils::http::login> &credentials)
```
What issue do you want to address?

xiphon commented on src/rpc/bootstrap_daemon.h:
```diff
@@ -0,0 +1,184 @@
+    }
+
+  private:
+    bool set_server(const std::string &address, const boost::optional<epee::net_utils::http::login> &credentials) noexcept
```
What issue do you want to address?

xiphon commented:
Updated. Refactored the code a lot

* Moved all the peer list stuff to p2p `node_server` and `peerlist_manager` classes.
* Implemented `get_public_nodes` and `get_random_public_node` in p2p `node_server` class, 
```c++
void node_server::get_public_nodes(PeerType type, std::vector<peerlist_entry>& public_nodes);
boost::optional<std::string> node_server::get_random_public_node();
```
* `O(1)` instead of `O(N)` to fetch public nodes from the peer list.  No more full list scans. Using hashed non unique index.
```c++
void peerlist_manager::get_public_nodes(PeerType type, std::vector<peerlist_entry>& public_nodes);
```
* Separate `bootstrap_daemon.cpp`
* Fixed `noexcept` declarations
* Relying on http client connection state (`is_connected()`, `disconnect()`)
* Fetching current bootstrap daemon address from underlying http client (`get_host()`, `get_port()`)
* Reusing `invoke_http_json_rpc`
* `boost::optional` return values
* Dropped `nodetool::node_server<...> *m_p2p;` from `bootstrap_daemon` class

fuwa0529 commented:
> > Can you prevent these two flags from being used simultaneously?
> 
> Is not related to this PR

On a second thought, this free-rider problem probably isn't too bad in this case, the market will likely work it out. Ignore my concern.

xiphon commented:
> On a second thought, this free-rider problem probably isn't too bad in this case, the market will likely work it out. Ignore my concern.

That is valid concern overall and it *might* be addressed later by altering '--no-sync' logic. For example, '--no-sync' node might periodically connect to p2p and disconnect right after refreshing the peer list and fetching current blockchain state (target height, etc.).

xiphon commented:
Updated the code once again. Stripped out most of the yet unneeded optimizations (mostly peer list fetching stuff), can introduce them later if it will become a bottleneck, though i highly doubt it will ever happen.
